Dockerfile

```diff
@@ -1,15 +1,45 @@
+# LAYER builder — — — — — — — — — — — — — — — — — — — — — — —
 FROM golang:1.12-alpine AS builder
 ADD go.mod go.sum /app/
 WORKDIR /app/
 RUN apk --no-cache add git upx
 RUN go mod download
 ADD . /app/
 RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -ldflags "-X main.version=$(git describe --tags --abbrev=0)" -a -o /trivy cmd/trivy/main.go
-RUN upx --lzma --best /trivy
 
-FROM alpine:3.9
-RUN apk --no-cache add ca-certificates git
+
+# LAYER compress — — — — — — — — — — — — — — — — — — — — — — —
+FROM alpine:3.9 AS compress
 COPY --from=builder /trivy /usr/local/bin/trivy
-RUN chmod +x /usr/local/bin/trivy
 
-ENTRYPOINT ["trivy"]
+RUN set -eux && apk --update --no-cache add \
+    upx && \
+    upx --lzma --best /usr/local/bin/trivy && \
+# unit test
+    upx -t /usr/local/bin/trivy && \
+    trivy --version
+
+
+# LAYER final — — — — — — — — — — — — — — — — — — — — — — — —
+FROM alpine:3.9 AS final
+COPY --from=compress /usr/local/bin/trivy /usr/local/bin/trivy
+
+RUN set -eux && apk --update --no-cache add \
+    ca-certificates git && \
+    chmod +x /usr/local/bin/trivy
+
+# best practice credit: https://github.com/opencontainers/image-spec/blob/master/annotations.md
+LABEL org.opencontainers.image.authors="Author"                                         \
+      org.opencontainers.image.vendors="Vendor"                                         \
+      org.opencontainers.image.created="${CREATED_DATE}"                                \
+      org.opencontainers.image.revision="${SOURCE_COMMIT}"                              \
+      org.opencontainers.image.title="Trivy"                                            \
+      org.opencontainers.image.description="Docker image for Trivy"                     \
+      org.opencontainers.image.url="https://hub.docker.com/r/0o0o/0o0o/tags/"           \
+      org.opencontainers.image.source="https://github.com/0o0o/0o0o"                    \
+      org.opencontainers.image.licenses="https://github.com/0o0o/0o0o/LICENSE.md"       \
+      org.example.image.user="root"                                                     \
+      org.example.image.alpineversion="3.9"                                             \
+      org.example.image.schemaversion="1.0"
+
+CMD [ "trivy" ]
```

Comment:
Referring to the previous discussion about this line (https://github.com/aquasecurity/trivy/pull/99#discussion_r307968720) - the change from calling `trivy` from the ENTRYPOINT to the CMD does not really make sense...

The first thing I did with the image after building it was to kick the tires:
```
docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock trivy
```
```
2020-01-06T16:02:08.635Z        ERROR   trivy requires at least 1 argument or --input option
NAME:
  trivy - A simple and comprehensive vulnerability scanner for containers
USAGE:
...
```
That's what I expected - I'd need to pass some arguments. Happily, `trivy` printed usage information telling me that I needed to state the image as the first positional argument - I can now get to work:
```
docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock trivy ubuntu:18.04
```
```
docker: Error response from daemon: OCI runtime create failed: container_linux.go:345: starting container process caused "exec: \"ubuntu:18.04\": executable file not found in $PATH": unknown.
```
This is not what I expected to happen, given the usage information.

Setting the command to the application like this only makes sense if the application is a service. As `trivy` is not, the only thing that can ever happen is that it is envoked without any arguments and fails... that's not the most useful thing in the world! It would therefore be better if changed back to ENTRYPOINT. 

> I really dislike it when people do this. It prvents you from doing `docker run --rm -it trivy/trivy bash`

@SuperSandro2000 if you want to hack around in a container when the entrypoint has been set to something that's not a shell, the way to do it is to override the entrypoint: `docker run --rm -it --entrypoint=bash trivy/trivy`.

Reply:
> This is not what I expected to happen

And I expect it to run bash cause all official images do function like this and adding ``--entrypoint`` works but it is not nice.

Reply:
I would love to see something like https://github.com/nodejs/docker-node/blob/master/10/buster/docker-entrypoint.sh .
Adds 116 bytes and fixes my and your problem.

Reply:
I've not reviewed what all official images do but I would comment that running a shell inside such a container is perhaps not most people's the main use case. It is interesting to note that "distroless" images are seemingly becoming popular now (e.g. those by Google https://github.com/GoogleContainerTools/distroless), which do not support such use case at all.

I wouldn't mind a more exotic entrypoint as you suggested.

Reply:
I fixed this: https://github.com/aquasecurity/trivy/pull/99/commits/a4b74f494f38933ea597978cc7087245c094498a#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557L45